### PR TITLE
ffi: Add method to do wildcard match on encoded variables; Some clean-up.

### PR DESCRIPTION
**Pull request by kirkrodrigues**
# Description

This PR mainly introduces a method for performing wildcard matches against encoded variables and does some minor cleanup of the ffi library.

# Validation performed

Validated old and new unit tests pass.

### CODE REVIEW

File: components/core/src/ffi/encoding_methods.hpp [hunk 1]
```diff
@@ -249,6 +249,35 @@ namespace ffi {
...
+     * Checks whether the given wildcard strings match the given encoded
+     * variables (from a message). Specifically, let {w in W} be the set of
+     * wildcard strings and {e in E} be the set of encoded variables. This
+     * method will return true only if:
+     * (1) Each unique `w` matches a unique `e`.
+     * (2) When (1) is true, the order of elements in both W and E is unchanged.
+     * @tparam encoded_variable_t Type of the encoded variable
+     * @param logtype The message's logtype
+     * @param encoded_vars The message's encoded variables
+     * @param encoded_vars_length
```
Comment (haiqi96) [inline components/core/src/ffi/encoding_methods.hpp:263]: can we be clear on what length mean? does it mean number of encoded_vars or size of the encoded_var buffer?

File: components/core/src/ffi/encoding_methods.hpp [hunk 2]
```diff
@@ -249,6 +249,35 @@ namespace ffi {
...
+     * no redundant wildcards (e.g. "**") nor unnecessary escape characters
+     * (e.g. "\").
+     * @return Whether the wildcard strings match the encoded variables
+     */
+    template <typename encoded_variable_t>
+    bool wildcard_match_encoded_vars (
+            std::string_view logtype,
+            encoded_variable_t* encoded_vars,
+            int encoded_vars_length,
+            std::string_view wildcard_var_placeholders,
```
Comment (haiqi96) [inline components/core/src/ffi/encoding_methods.hpp:278]: I am thinking what changes will be required for the interface of this function when we merge ffi into clp. will we have two similar fiunctions with different interface or will we sort of combine the flows?

File: components/core/src/ffi/encoding_methods.tpp [hunk 3]
```diff
@@ -464,6 +464,60 @@ namespace ffi {
...
+
+    template <typename encoded_variable_t>
+    bool wildcard_match_encoded_vars (
+            std::string_view logtype,
+            encoded_variable_t* encoded_vars,
+            int encoded_vars_length,
+            std::string_view wildcard_var_placeholders,
+            const std::vector<std::string_view>& wildcard_var_queries
+    ) {
+        auto wildcard_var_queries_len = wildcard_var_queries.size();
```
Comment (haiqi96) [inline components/core/src/ffi/encoding_methods.tpp:476]: would be nice to check that wildcard_var_placeholders.length() = wildcard_var_queries.size()?

File: components/core/src/ffi/encoding_methods.hpp (hunk 1 above)
Comment (kirkrodrigues) [reply components/core/src/ffi/encoding_methods.hpp]: Clarified

File: components/core/src/ffi/encoding_methods.hpp (hunk 2 above)
Comment (kirkrodrigues) [reply components/core/src/ffi/encoding_methods.hpp:282]: We will likely introduce another interface for CLP's internal use. I added a comment explaining why this interface was chosen.

File: components/core/src/ffi/encoding_methods.tpp (hunk 3 above)
Comment (kirkrodrigues) [reply components/core/src/ffi/encoding_methods.tpp:486]: Good idea! Added some parameter validation.

**Review by haiqi96 — APPROVED** [review]

**Pull request merged**